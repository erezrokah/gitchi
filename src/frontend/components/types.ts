interface Author {
  avatarUrl: string;
  login: string;
}

interface User extends Author {
  name: string;
}

interface Comment {
  id: string;
  bodyText: string;
  createdAt: Date;
  author: Author;
  inReplyToId?: number;
  path?: string;
}

interface Channel {
  key: string;
  title: string;
  isReview: boolean;
  comments: Comment[];
}

interface Pr {
  id: string;
  bodyText: string;
  title: string;
  author: Author;
  channels: Channel[];
}
