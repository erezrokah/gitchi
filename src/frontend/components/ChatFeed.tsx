import * as React from 'react';
import { createRef, useEffect, forwardRef, Fragment } from 'react';
import { Feed, Segment, Icon, Image } from 'semantic-ui-react';
import styled from 'styled-components';
import { formatRelative } from 'date-fns';

const StyledSegment = styled(Segment)`
  height: 500px;
  width: 500px;
  overflow: auto;
`;

interface Props {
  onDeleteMessage: (id: string) => void;
  comments: Comment[];
}

interface FeedEventProps {
  id: string;
  avatarUrl: string;
  login: string;
  bodyText: string;
  createdAt: Date;
  canDelete: boolean;
  onDeleteMessage: (id: string) => void;
}

const FeedEvent = (props: FeedEventProps) => {
  const {
    id,
    avatarUrl,
    login,
    bodyText,
    createdAt,
    canDelete,
    onDeleteMessage,
  } = props;

  return (
    <Fragment>
      <Feed.Label>
        <Image src={avatarUrl} />
      </Feed.Label>
      <Feed.Content>
        <Feed.Summary>
          <Feed.User>{login}</Feed.User>
          <Feed.Extra text>{bodyText}</Feed.Extra>
          <Feed.Date>{formatRelative(createdAt, new Date())}</Feed.Date>
        </Feed.Summary>
        {canDelete ? (
          <Feed.Meta>
            <Feed.Like>
              <Icon name="delete" onClick={() => onDeleteMessage(id)} />
            </Feed.Like>
          </Feed.Meta>
        ) : null}
      </Feed.Content>
    </Fragment>
  );
};

export const ChatFeed = (props: Props) => {
  const lastCommentRef = createRef<HTMLDivElement>();

  useEffect(() => {
    if (lastCommentRef.current) {
      lastCommentRef.current.scrollIntoView({ behavior: 'smooth' });
    }
  }, [lastCommentRef]);

  const events = props.comments.map(
    ({ bodyText, id, author, createdAt, canDelete }, index) => {
      if (props.comments.length - 1 === index) {
        function feedEventWithRef(
          props: FeedEventProps,
          ref: React.Ref<HTMLDivElement>,
        ) {
          return (
            <div ref={ref} className="event">
              <FeedEvent {...props} />
            </div>
          );
        }

        const FeedEventWithRef = forwardRef<HTMLDivElement, FeedEventProps>(
          feedEventWithRef,
        );

        return (
          <FeedEventWithRef
            key={id}
            ref={lastCommentRef}
            id={id}
            bodyText={bodyText}
            createdAt={createdAt}
            avatarUrl={author.avatarUrl}
            login={author.login}
            canDelete={canDelete}
            onDeleteMessage={props.onDeleteMessage}
          />
        );
      } else {
        return (
          <div key={id} className="event">
            <FeedEvent
              id={id}
              bodyText={bodyText}
              createdAt={createdAt}
              avatarUrl={author.avatarUrl}
              login={author.login}
              canDelete={canDelete}
              onDeleteMessage={props.onDeleteMessage}
            />
          </div>
        );
      }
    },
  );

  return (
    <StyledSegment attached="bottom">
      <Feed>{events}</Feed>
    </StyledSegment>
  );
};
