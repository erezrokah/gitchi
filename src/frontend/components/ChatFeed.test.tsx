import * as React from 'react';
import { render } from '@testing-library/react';
import { ChatFeed } from './ChatFeed';
import { User } from './types';

const scrollIntoView = jest.fn();
window.HTMLElement.prototype.scrollIntoView = scrollIntoView;

describe('ChatFeed', () => {
  const user: User = {
    name: 'name',
    avatarUrl: 'avatarUrl',
    login: 'login',
  };

  const author = {
    avatarUrl: 'authorAvatarUrl',
    login: 'authorLogin',
  };

  const props = {
    onDeleteMessage: jest.fn(),
    comments: [
      {
        id: '100',
        bodyText: 'comment 100',
        createdAt: new Date(100),
        author,
        canDelete: false,
      },
      {
        id: '200',
        bodyText: 'comment 200',
        createdAt: new Date(200),
        author: user,
        canDelete: true,
      },
    ],
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should match snapshot', () => {
    const { asFragment } = render(<ChatFeed {...props} />);
    expect(asFragment()).toMatchSnapshot();
    expect(scrollIntoView).toHaveBeenCalledTimes(1);
  });

  test('should match snapshot with empty comments array', () => {
    const { asFragment } = render(<ChatFeed {...props} comments={[]} />);
    expect(asFragment()).toMatchSnapshot();
    expect(scrollIntoView).toHaveBeenCalledTimes(0);
  });
});
