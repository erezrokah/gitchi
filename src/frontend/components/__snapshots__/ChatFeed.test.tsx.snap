// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatFeed should match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  height: 500px;
  width: 500px;
  overflow: auto;
}

<div
    class="ui bottom attached segment c0"
  >
    <div
      class="ui feed"
    >
      <div
        class="event"
      >
        <div
          class="label"
        >
          <img
            class="ui image"
            src="authorAvatarUrl"
          />
        </div>
        <div
          class="content"
        >
          <div
            class="summary"
          >
            <a
              class="user"
            >
              authorLogin
            </a>
            <div
              class="text extra"
            >
              comment 100
            </div>
            <div
              class="date"
            >
              01/01/1970
            </div>
          </div>
        </div>
      </div>
      <div
        class="event"
      >
        <div
          class="label"
        >
          <img
            class="ui image"
            src="avatarUrl"
          />
        </div>
        <div
          class="content"
        >
          <div
            class="summary"
          >
            <a
              class="user"
            >
              login
            </a>
            <div
              class="text extra"
            >
              comment 200
            </div>
            <div
              class="date"
            >
              01/01/1970
            </div>
          </div>
          <div
            class="meta"
          >
            <a
              class="like"
            >
              <i
                aria-hidden="true"
                class="delete icon"
                data-testid="delete-comment-200"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ChatFeed should match snapshot with empty comments array 1`] = `
<DocumentFragment>
  .c0 {
  height: 500px;
  width: 500px;
  overflow: auto;
}

<div
    class="ui bottom attached segment c0"
  >
    <div
      class="ui feed"
    />
  </div>
</DocumentFragment>
`;
